// @ngrx/eslint-plugin/src/rules/store/prefix-selectors-with-select.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.prefixSelectorsWithSelectSuggest = exports.prefixSelectorsWithSelect = void 0;
const path = __importStar(require("path"));
const rule_creator_1 = require("../../rule-creator");
const utils_1 = require("../../utils");
exports.prefixSelectorsWithSelect = 'prefixSelectorsWithSelect';
exports.prefixSelectorsWithSelectSuggest = 'prefixSelectorsWithSelectSuggest';
exports.default = (0, rule_creator_1.createRule)({
    name: path.parse(__filename).name,
    meta: {
        type: 'suggestion',
        hasSuggestions: true,
        docs: {
            description: 'The selector should start with "select", for example "selectEntity".',
            ngrxModule: 'store',
        },
        schema: [],
        messages: {
            [exports.prefixSelectorsWithSelect]: 'The selector should start with "select".',
            [exports.prefixSelectorsWithSelectSuggest]: 'Prefix the selector with "select": `{{ name }}`.',
        },
    },
    defaultOptions: [],
    create: (context) => {
        function reportIfInvalid(name, node) {
            // Name starts with select and
            // the first character after select is an uppercase ASCII letter, _, or $
            const isValid = name.startsWith('select') &&
                name.length > 'select'.length &&
                /^[A-Z_$]/.test(name.slice('select'.length));
            if (!isValid) {
                const suggestedName = getSuggestedName(name);
                context.report({
                    node,
                    loc: {
                        start: node.loc.start,
                        end: {
                            line: node.loc.start.line,
                            column: node.loc.start.column + name.length,
                        },
                    },
                    messageId: exports.prefixSelectorsWithSelect,
                    suggest: [
                        {
                            messageId: exports.prefixSelectorsWithSelectSuggest,
                            data: { name: suggestedName },
                            fix: (fixer) => {
                                const parent = node.parent;
                                const sourceCode = context.getSourceCode();
                                // Handle destructuring: { selectAll: allItems }
                                if (parent &&
                                    parent.type === 'Property' &&
                                    parent.value === node &&
                                    parent.parent &&
                                    parent.parent.type === 'ObjectPattern') {
                                    return fixer.replaceText(node, suggestedName);
                                }
                                // Handle simple variable declarator: const allItems = ...
                                if (parent &&
                                    parent.type === 'VariableDeclarator' &&
                                    parent.id.type === 'Identifier') {
                                    const typeAnnotation = parent.id.typeAnnotation
                                        ? sourceCode.getText(parent.id.typeAnnotation)
                                        : '';
                                    return fixer.replaceText(parent.id, `${suggestedName}${typeAnnotation}`);
                                }
                                // Fallback: just replace the identifier
                                return fixer.replaceText(node, suggestedName);
                            },
                        },
                    ],
                });
            }
        }
        function isSelectorFactoryCall(node) {
            const callee = node.callee;
            return (callee.type === 'Identifier' &&
                [
                    'createSelector',
                    'createFeatureSelector',
                    'createSelectorFactory',
                ].includes(callee.name));
        }
        function checkFunctionBody(name, node, body) {
            if (body.type === 'CallExpression' && isSelectorFactoryCall(body)) {
                reportIfInvalid(name, node);
            }
            if (body.type === 'BlockStatement') {
                for (const stmt of body.body) {
                    if (stmt.type === 'ReturnStatement' &&
                        stmt.argument &&
                        stmt.argument.type === 'CallExpression' &&
                        isSelectorFactoryCall(stmt.argument)) {
                        reportIfInvalid(name, node);
                    }
                }
            }
        }
        return {
            VariableDeclarator(node) {
                const { id, init } = node;
                const isSelectorSource = init?.type === 'CallExpression' &&
                    ((init.callee.type === 'Identifier' &&
                        init.callee.name === 'getSelectors') ||
                        (init.callee.type === 'MemberExpression' &&
                            init.callee.property.type === 'Identifier' &&
                            init.callee.property.name === 'getSelectors'));
                if (id.type === 'ObjectPattern' && isSelectorSource) {
                    for (const prop of id.properties) {
                        if (prop.type === 'Property' && prop.value.type === 'Identifier') {
                            reportIfInvalid(prop.value.name, prop.value);
                        }
                    }
                    return;
                }
                if (id.type === 'Identifier') {
                    const typeName = node.id.typeAnnotation?.typeAnnotation.type === 'TSTypeReference' &&
                        node.id.typeAnnotation.typeAnnotation.typeName.type === 'Identifier'
                        ? node.id.typeAnnotation.typeAnnotation.typeName.name
                        : null;
                    const hasSelectorType = typeName !== null && /Selector$/.test(typeName);
                    const isSelectorCall = init?.type === 'CallExpression' && isSelectorFactoryCall(init);
                    const isArrowFunction = init?.type === 'ArrowFunctionExpression' &&
                        init.body &&
                        (init.body.type === 'CallExpression' ||
                            init.body.type === 'BlockStatement');
                    const isFunctionExpression = init?.type === 'FunctionExpression' &&
                        init.body &&
                        init.body.type === 'BlockStatement';
                    if (hasSelectorType || isSelectorCall) {
                        reportIfInvalid(id.name, id);
                    }
                    else if (isArrowFunction || isFunctionExpression) {
                        checkFunctionBody(id.name, id, init.body);
                    }
                }
            },
        };
    },
});
function getSuggestedName(name) {
    const selectWord = 'select';
    if (name.startsWith(selectWord)) {
        const rest = name.slice(selectWord.length);
        if (rest.length === 0) {
            return 'selectSelect';
        }
        if (/^[A-Z_]+$/.test(rest)) {
            return `${selectWord}${rest}`;
        }
        return `${selectWord}${(0, utils_1.capitalize)(rest)}`;
    }
    if (/^get([^a-z].+)/.test(name)) {
        const rest = name.slice(3);
        return `${selectWord}${(0, utils_1.capitalize)(rest)}`;
    }
    if (/^[A-Z_]+$/.test(name)) {
        return `${selectWord}${name}`;
    }
    return `${selectWord}${(0, utils_1.capitalize)(name)}`;
}
