// @ngrx/eslint-plugin/src/configs/effects.js
"use strict";
/**
 * DO NOT EDIT
 * This file is generated
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = (plugin, parser) => [
    {
        name: 'ngrx/base',
        languageOptions: {
            parser,
            sourceType: 'module',
        },
        plugins: {
            '@ngrx': plugin,
        },
    },
    {
        name: 'ngrx/effects',
        languageOptions: {
            parser,
            parserOptions: {
                ecmaVersion: 2020,
                sourceType: 'module',
                project: './tsconfig.json',
            },
        },
        rules: {
            '@ngrx/avoid-cyclic-effects': 'error',
            '@ngrx/no-dispatch-in-effects': 'error',
            '@ngrx/no-effects-in-providers': 'error',
            '@ngrx/no-multiple-actions-in-effects': 'error',
            '@ngrx/prefer-action-creator-in-of-type': 'error',
            '@ngrx/prefer-effect-callback-in-block-statement': 'error',
            '@ngrx/use-effects-lifecycle-interface': 'error',
        },
    },
];
