// @ngrx/eslint-plugin/v9/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configs = exports.meta = void 0;
const typescript_eslint_1 = require("typescript-eslint");
const rules_1 = require("../src/rules");
const package_json_1 = require("../package.json");
const all_1 = __importDefault(require("../src/configs/all"));
const store_1 = __importDefault(require("../src/configs/store"));
const effects_1 = __importDefault(require("../src/configs/effects"));
const component_store_1 = __importDefault(require("../src/configs/component-store"));
const operators_1 = __importDefault(require("../src/configs/operators"));
const signals_1 = __importDefault(require("../src/configs/signals"));
const meta = { name: package_json_1.name, version: package_json_1.version };
exports.meta = meta;
const tsPlugin = {
    rules: rules_1.rules,
};
const configs = {
    all: (0, all_1.default)(tsPlugin, typescript_eslint_1.parser),
    store: (0, store_1.default)(tsPlugin, typescript_eslint_1.parser),
    effects: (0, effects_1.default)(tsPlugin, typescript_eslint_1.parser),
    componentStore: (0, component_store_1.default)(tsPlugin, typescript_eslint_1.parser),
    operators: (0, operators_1.default)(tsPlugin, typescript_eslint_1.parser),
    signals: (0, signals_1.default)(tsPlugin, typescript_eslint_1.parser),
};
exports.configs = configs;
/*
As the angular-eslint plugin we do both a default and named exports to allow people to use this package from
both CJS and ESM in very natural ways.
*/
exports.default = {
    meta,
    configs,
};
