// @ngrx/eslint-plugin/scripts/generate-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path_1 = require("path");
const prettier_1 = require("prettier");
const rules_1 = require("../src/utils/helper-functions/rules");
(async () => {
    const prettierConfig = await (0, prettier_1.resolveConfig)(__dirname);
    const RULE_MODULE = '@ngrx';
    const CONFIG_DIRECTORY = './modules/eslint-plugin/src/configs/';
    writeConfig('all', (_rule) => true);
    writeConfig('store', (rule) => rule.meta.docs?.ngrxModule === 'store');
    writeConfig('effects', (rule) => rule.meta.docs?.ngrxModule === 'effects');
    writeConfig('component-store', (rule) => rule.meta.docs?.ngrxModule === 'component-store');
    writeConfig('operators', (rule) => rule.meta.docs?.ngrxModule === 'operators');
    writeConfig('signals', (rule) => rule.meta.docs?.ngrxModule === 'signals');
    async function writeConfig(configName, predicate) {
        const rulesForConfig = Object.entries(rules_1.rulesForGenerate).filter(([_, rule]) => predicate(rule));
        const configRules = rulesForConfig.reduce((rules, [ruleName, _rule]) => {
            rules[`${RULE_MODULE}/${ruleName}`] = 'error';
            return rules;
        }, {});
        const requireParserOptions = rulesForConfig.some(([_, rule]) => rule.meta.docs?.requiresTypeChecking)
            ? {
                ecmaVersion: 2020,
                sourceType: 'module',
                project: './tsconfig.json',
            }
            : null;
        const tsCode = `
      /**
     * DO NOT EDIT
     * This file is generated
     */

      import type { TSESLint } from '@typescript-eslint/utils';

      export default (
        plugin: TSESLint.FlatConfig.Plugin,
        parser: TSESLint.FlatConfig.Parser,
      ): TSESLint.FlatConfig.ConfigArray => [
        {
          name: 'ngrx/base',
          languageOptions: {
            parser,
            sourceType: 'module',
          },
          plugins: {
            '@ngrx': plugin,
          },
        },
        {
          name: 'ngrx/${configName}',
          languageOptions: {
            parser,
            ${requireParserOptions
            ? `parserOptions: ${JSON.stringify(requireParserOptions, null, 2)},`
            : ''}
          },
          rules: ${JSON.stringify(configRules, null, 2)}
        },
      ];`;
        const tsConfigFormatted = await (0, prettier_1.format)(tsCode, {
            parser: 'typescript',
            ...prettierConfig,
        });
        (0, fs_1.writeFileSync)((0, path_1.join)(CONFIG_DIRECTORY, `${configName}.ts`), tsConfigFormatted);
        const jsonConfig = {
            parser: '@typescript-eslint/parser',
            plugins: ['@ngrx'],
            rules: configRules,
        };
        if (requireParserOptions) {
            jsonConfig.parserOptions = {
                ecmaVersion: 2020,
                sourceType: 'module',
                project: './tsconfig.json',
            };
        }
        const jsonConfigFormatted = await (0, prettier_1.format)(JSON.stringify(jsonConfig, null, 2), {
            parser: 'json',
            ...prettierConfig,
        });
        (0, fs_1.writeFileSync)((0, path_1.join)(CONFIG_DIRECTORY, `${configName}.json`), jsonConfigFormatted);
    }
})();
